src/lazy/encoder/mod.rs

```diff
@@ -0,0 +1,565 @@
+    fn end(mut self) -> IonResult<()> {
+        let space_between_top_level_values =
+            self.whitespace_config().space_between_top_level_values;
+        write!(self.output(), "]{space_between_top_level_values}")?;
+        self.has_been_closed = true;
+        Ok(())
+    }
+
+    fn output(&mut self) -> &mut W {
+        &mut self.writer.output
+    }
+
+    fn whitespace_config(&self) -> &WhitespaceConfig {
+        self.writer.whitespace_config
+    }
+
+    #[inline]
+    fn value_writer(&mut self) -> TextValueWriter_1_0<'_, W> {
+        TextValueWriter_1_0 {
+            writer: self.writer,
+            depth: self.depth,
+        }
+    }
+
+    #[inline]
+    fn annotated_value_writer(&mut self) -> TextAnnotatedValueWriter_1_0<'_, W> {
+        TextAnnotatedValueWriter_1_0 {
+            value_writer: self.value_writer(),
+        }
+    }
+}
+
+impl<'a, W: Write> Drop for TextListWriter_1_0<'a, W> {
+    fn drop(&mut self) {
+        // If the user didn't call `end`, the closing delimiter was not written to output.
+        // It's too late to call it here because we can't return a Result.
+        if !self.has_been_closed {
+            panic!("List writer was dropped without calling `end()`.");
+        }
+    }
+}
+
+impl<'a, W: Write, E: LazyEncoder<W>> SequenceWriter<'a, W, E> for TextListWriter_1_0<'a, W> {
+    delegate! {
+        to self {
+            fn write<V: WriteAsIon>(&mut self, value: V) -> IonResult<&mut Self>;
+            fn end(self) -> IonResult<()>;
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::lazy::encoder::annotate::Annotate;
+    use crate::lazy::encoder::LazyRawTextWriter_1_0;
+    use crate::symbol_ref::AsSymbolRef;
+    use crate::{Element, IonData, IonResult, Timestamp};
+
+    fn writer_test(
+        expected: &str,
+        mut test: impl FnMut(&mut LazyRawTextWriter_1_0<&mut Vec<u8>>) -> IonResult<()>,
+    ) -> IonResult<()> {
+        let expected = Element::read_all(expected)?;
+        let mut buffer = Vec::new();
+        let mut writer = LazyRawTextWriter_1_0::new(&mut buffer);
+        test(&mut writer)?;
+        writer.flush()?;
+        let actual = Element::read_all(buffer)?;
+        assert!(IonData::eq(&expected, &actual));
+        Ok(())
+    }
+
+    #[test]
+    fn write_scalars() -> IonResult<()> {
+        let expected = r#"
+            1
+            false
+            3e0
+            "foo"
+            bar
+            2023-11-09T
+            {{4AEA6g==}}
+            [1, 2, 3]
+        "#;
+        let test = |writer: &mut LazyRawTextWriter_1_0<&mut Vec<u8>>| {
+            writer
+                .write(1)?
+                .write(false)?
+                .write(3f32)?
+                .write("foo")?
+                .write("bar".as_symbol_ref())?
+                .write(Timestamp::with_ymd(2023, 11, 9).build()?)?
+                .write([0xE0u8, 0x01, 0x00, 0xEA])?
+                .write([1, 2, 3])?;
```

Comment:
🗺️ There are two ways to write a list: if you have a type that represents the list and your type implements `WriteAsIon[Value]`, you can pass it to `.write(...)` as is done here. For a streaming list writer that can emit heterogeneous values, you can use `ListWriter`.

If `WriteAsIon[Value]` is implemented, it would be implemented using `ListWriter`.